optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolver.java

```diff
@@ -163,15 +162,22 @@ public boolean isEveryProblemFactChangeProcessed() {
     }
 
     public void setMonitorTagMap(Map<String, String> monitorTagMap) {
-        monitoringTags = ObjectUtils.defaultIfNull(monitorTagMap, Collections.<String, String> emptyMap())
+        Tags monitoringTags = ObjectUtils.defaultIfNull(monitorTagMap, Collections.<String, String> emptyMap())
                 .entrySet().stream().map(entry -> Tags.of(entry.getKey(), entry.getValue()))
                 .reduce(Tags.empty(), Tags::and);
-        solverScope.setMonitoringTags(monitoringTags);
         unregisterMetrics();
+        solverScope.setMonitoringTags(monitoringTags);
         registerMetrics();
```

Comment:
This is not particularly clean. 

First you unregister tags here by querying solver scope for them.
Then you set some other tags to the solver scope.
And then you one again read them to register them with this class.

I'd argue that all these operations should be happening from one place, and the data should be encapsulated there.
The more you "leak" internals to other classes, the harder it becomes over time to figure out who does what with them where.

Reply:
solverScope is the single source of tags (since these tags also need to be accessed in phase listeners + other places in the code). The metrics in `Solver` are unique in that they are created before `solving` (for performance reasons, Micrometer recommends storing meters in fields if possible), so they are the only meters that can change after they are created. These operations probably won't be needed if we make the fields local to the `solve` call (at a very sight performance penalty).

Reply:
Doing it in `solve` probably isn't such a bad idea, provided that these operations are really cheap.